example/service/s3/listObjects/listObjects.go

```diff
@@ -0,0 +1,129 @@
+	flag.StringVar(&objectDelimiter, "delimiter", "/",
+		"The optional `object key delimiter` used by S3 List objects to group object keys.")
+}
+
+// Lists all objects in a bucket using pagination
+//
+// Usage:
+// listObjects <bucket>
+func main() {
+	flag.Parse()
+	if len(bucketName) == 0 {
+		flag.PrintDefaults()
+		log.Fatalf("invalid parameters, bucket name required")
+	}
+
+	// Load the SDK's configuration from environment and shared config, and
+	// create the client with this.
+	cfg, err := config.LoadDefaultConfig()
+	if err != nil {
+		log.Fatalf("failed to load SDK configuration, %v", err)
+	}
+
+	client := s3.NewFromConfig(cfg)
+
+	// Set the parameters based on the CLI flag inputs.
+	params := &s3.ListObjectsV2Input{
+		Bucket: &bucketName,
+	}
+	if len(objectPrefix) != 0 {
+		params.Prefix = &objectPrefix
+	}
+	if len(objectDelimiter) != 0 {
+		params.Delimiter = &objectDelimiter
+	}
+
+	// TODO replace this with the code generate paginator when available
+	// s3.NewListObjectsV2Paginator()
+	p := NewS3ListObjectsV2Paginator(client, params)
+
+	// Iterate through the S3 object pages, printing each object returned.
+	var i int
+	log.Println("Objects:")
+	for p.HasMorePages() {
+		i++
+
+		page, err := p.NextPage(context.TODO())
+		if err != nil {
+			log.Fatalf("failed to get page %v, %v", i, err)
+		}
+
+		// Log the objects found
+		for _, obj := range page.Contents {
+			fmt.Println("Object:", *obj.Key)
+		}
+	}
+}
+
+// S3ListObjectsV2APIClient provides interface for the S3 API client
+// ListObjectsV2 operation call.
+type S3ListObjectsV2APIClient interface {
+	ListObjectsV2(context.Context, *s3.ListObjectsV2Input, ...func(*s3.Options)) (*s3.ListObjectsV2Output, error)
+}
+
+// S3ListObjectsV2Paginator provides the paginator to paginate S3 ListObjectsV2
+// response pages.
+type S3ListObjectsV2Paginator struct {
+	client S3ListObjectsV2APIClient
+	params *s3.ListObjectsV2Input
+
+	nextToken *string
+	firstPage bool
+}
+
+// NewS3ListObjectsV2Paginator initializes a new S3 ListObjectsV2 Paginator for
+// paginating the ListObjectsV2 respones.
+func NewS3ListObjectsV2Paginator(client S3ListObjectsV2APIClient, params *s3.ListObjectsV2Input) *S3ListObjectsV2Paginator {
+	return &S3ListObjectsV2Paginator{
+		client:    client,
+		params:    params,
+		firstPage: true,
+	}
+}
+
+// HasMorePages returns true if there are more pages or if the first page has
+// not been retrieved yet.
+func (p *S3ListObjectsV2Paginator) HasMorePages() bool {
+	return p.firstPage || (p.nextToken != nil && len(*p.nextToken) != 0)
+}
+
+// NextPage attempts to retrieve the next page, or returns error if unable to.
+func (p *S3ListObjectsV2Paginator) NextPage(ctx context.Context) (
+	*s3.ListObjectsV2Output, error,
+) {
+	if !p.HasMorePages() {
+		return nil, fmt.Errorf("no more pages available")
+	}
+
+	params := *p.params
+	result, err := p.client.ListObjectsV2(ctx, &params)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
You don't have to take the value of `p.params` here, as you are sending the `&params` in the next step

If the intention is to use a copy of input; would it make sense for Paginator struct to not take a input value (rather than a pointer).

Reply:
The intention is to create a copy of `p.params` that will not mutate the original.

Taking a value for params at input instead of a pointer could be helpful, though it is asymmetric to the other API calls. I don't think we've discussed this yet with the design of the Paginators.

One benefit of taking a pointer as input, is that for operations that have no required members you can pass in `nil` instead. If values were taken here that would be another asymmetric usage of an input vs API.